### ISSUE TITLE
History flows into titlebar w/ decorations set to transparent

### ISSUE THREAD

**Issue by rtgibbons** [jwilm/alacritty#1671]
Which operating system does the issue occur on? macOS 10.14

If on linux, are you using X11 or Wayland? n/a

With Transparent decorations, the history will flow into the titlebar area and under the buttons.

![image](https://user-images.githubusercontent.com/8000/47098990-fe669e00-d1f9-11e8-96c3-e6e5bed4fc29.png)

**Comment by chrisduerr** [comment]
How does this behave with other terminal emulators? One could argue that this is the expected behavior, since the decorations are truly transparent. I'd assume that instead of transparent decorations, the decorations are just embedded into the application with a flat background color?

Pinging @sodiumjoe since he's the expert for this.

**Comment by monouser7dig** [comment]
kitty does not draw into the title bar

**Comment by sodiumjoe** [comment]
This is expected behavior, you'll have to add padding yourself. I also have a separate PR up for granular padding adjustments: #1290

**Comment by chrisduerr** [comment]
@sodiumjoe What motivated you to go for this choice? How does iTerm/Terminal.app behave? If no application on macOS makes use of the space below the transparent titlebar, it would probably be good to automatically follow that example without the need of tweaking the config.

**Comment by rtgibbons** [comment]
That's what is a bit weird, Most still use the title bar space for something else. iTerm2 Nightly is using it for the window title and optional tab bar, chrome and firefox put tabs up there. So the window content never really has a chance to flow up there.

When open a new window the initial prompt is properly spaced. If I add a padding.y of 22 to the config the initial prompt is now lower, but the history doesn't flow into the title bar.

**Comment by rtgibbons** [comment]
Ex. of buttonless after typing clear, it auto adjust for what /would/ be the Titlebar

![image](https://user-images.githubusercontent.com/8000/47118053-c11b0400-d22b-11e8-995c-657e19d14d04.png)

**Comment by sodiumjoe** [comment]
@chrisduerr iirc, I couldn't replicate what iterm was doing exactly with winit's api, i.e. without more control over the window. I also couldn't find a clean way to figure out how tall a titlebar would be across all versions of macos.

I agree that keeping the space for where the titlebar is normally would be ideal, but since I don't use `transparent`, it wasn't a priority for me to try to implement it. I can take another look when I get some time, though.

**Comment by sodiumjoe** [comment]
@rtgibbons that's with no padding? That's strange, this is what I get with `buttonless` and no padding:

![image](https://user-images.githubusercontent.com/845857/47118302-bce6ea80-d21b-11e8-9129-e6f3d219ab28.png)

**Comment by rtgibbons** [comment]
That gap is only when I clear terminal or open a new window, hoping it might lead to a way to detect the titlebar height. Here it is when stuff open 

![image](https://user-images.githubusercontent.com/8000/47118584-4bb03300-d22d-11e8-914c-ac3fa364dd87.png)

**Comment by sodiumjoe** [comment]
🤔 that's really strange, and I can't reproduce that. I wonder if that's an issue with where the excess padding goes when the grid doesn't line up with the window dimensions tho. (https://github.com/jwilm/alacritty/pull/1190) I mention it because it looks like there's some extra space at the bottom of the in your last screenshot.

**Comment by rtgibbons** [comment]
Yeah, i noticed that also, but that first screenshot showing it was on a new window creation, which should have the grid line up with the window (it at least appears that way when running `yes`)

**Comment by rtgibbons** [comment]
Testing gif showing the characteristics 
![out](https://user-images.githubusercontent.com/8000/47119393-515b4800-d230-11e8-9647-43b3c7069ff0.gif)

**Comment by sodiumjoe** [comment]
that looks like it may have to do with your shell initialization to me. Can you try with alacritty starting in `/bin/bash` with minimal configuration?

**Comment by rtgibbons** [comment]
Good catch, that's exactly what it is.

![image](https://user-images.githubusercontent.com/8000/47119586-0b52b400-d231-11e8-9d2b-ce82f8d21d61.png)

**Comment by rtgibbons** [comment]
I was digging into this a little.

So `with_titlebar_transparent` is doing exactly what it says, making the titlebar transparent. Then alacritty is using `with_fullsize_content_view` to bring the background color into the titlebar, but this is just moving the content view into the title bar which is now visible due to us setting it transparent. The downside to this is the actually content of the terminal is now in the titlebar. 

I think the fix for this would be fixed with winit, so that when Transparent decoration is set. `with_fullsize_content_view` would be set to false, and instead of `with_titlebar_transparent` we'd hopefully have winit function for `with_titlebar_color` which will give us a title bar with the same color. but keep the content out of the title bar. Then we could make the actual title optional at that point. But with buttons it makes sense to go this route, and then leave the current functionality which gives the added benefit of being able to drag the window around still.

**Comment by sodiumjoe** [comment]
@rtgibbons Yes, that sounds very familiar. At one point I did have it working with a transparent titlebar, such that the buttons were just floating. The methods I could find for setting the background of the titlebar were quite terrible, though. But that's not to discourage you from taking a stab, I knew basically nothing about rust or c or obj-c at the time (while now i know a tiny bit of those).

**Comment by rtgibbons** [comment]
Ha, I know just enough Obj-C to be dangerous and nothing really of Rust except how to fumble through reading it.

It appears that this titlebar color wasn't added afterall in winit, in favor of just transparent due to the API's able opened up. https://developer.apple.com/documentation/appkit/nswindow/1419167-titlebarappearstransparent?language=objc Which they state doesn't make sense to use without `NSFullSizeContentViewWindowMask` .. make sense.

It appears the proper way to do it is to then create a view at the top to build your own customer titlebar (i.e. one with tabs in it or toolbar i.e. Office 2016). What others have done if they just want to change the color is to hack around it with `addTitlebarAccessoryViewController` and have it be the full titlebar :/ Doesn't like there is any straight forward way of just making the titlebar the same color while keeping the spacing.

**Comment by sodiumjoe** [comment]
Yeah, that all sounds pretty familiar. I probably should've documented my findings better at the time 😂  Thanks for digging that up!

I agree with your conclusion. For the time being, the granular padding branch should work if you want to use `transparent`.

**Comment by lucasprag** [comment]
hey guys, great work on alacritty! I love it!
is there any workaround for now?
![Screen Shot 2019-03-14 at 11 33 24 AM](https://user-images.githubusercontent.com/2293178/54369922-1fb6a380-464d-11e9-8680-bfb0f7acc3cf.png)
thanks =)

**Comment by chrisduerr** [comment]
I believe the only choice for now is to use one of the other decoration options.

**Comment by lucasprag** [comment]
I see, I'm using `buttonless` for now, thanks

**Comment by chrisduerr** [comment]
Note that even with buttonless (see #2208) there are some issues with handling things this way on macOS. So if a default padding is added at the top, this needs to be done for both buttonless and transparent.

**Comment by lucasprag** [comment]
<img width="341" alt="Screen Shot 2019-03-24 at 11 22 25 PM" src="https://user-images.githubusercontent.com/2293178/54893348-b3794280-4e8b-11e9-8bb6-fdae47ad87f9.png">

Yes, bug now with macOS Mojave, Alacritty looks great without any decorations 🎉

**Comment by decrn** [comment]
Going off this thread, I have set an override for `window.padding.y`, but that obviously applies it to both top and bottom -- is there a way to set individual paddings?

**Comment by chrisduerr** [comment]
@decrn Currently there is not. Though it has been proposed.

I personally think that rather than having separate paddings for top/bottom, we should rather just try to solve this concrete problem. I don't think there are much other usecases for this.

**Comment by decrn** [comment]
I agree. Like others mentioned, perhaps we could incorporate some sort of title bar of sorts to occupy that space...
On Jun 7, 2019, 8:34 PM +0200, Christian Duerr <notifications@github.com>, wrote:

@decrn<https://github.com/decrn> Currently there is not. Though it has been proposed.

I personally think that rather than having separate paddings for top/bottom, we should rather just try to solve this concrete problem. I don't think there are much other usecases for this.

**Comment by sodiumjoe** [comment]
discussion of granular padding (and reasons for abandoning) are here: https://github.com/jwilm/alacritty/pull/1290

Some explanation of why we don't just add titlebar padding here: https://github.com/jwilm/alacritty/issues/1671#issuecomment-430801476

I think there have been lots of winit changes since I tried, so it might be worth revisiting.

It might also make sense to decide on a titlebar height instead of trying to figure out what the correct one should be for the current macOS version. I.e. use the tallest one, or use the latest version of macOS.

**Comment by chrisduerr** [comment]
> I think there have been lots of winit changes since I tried, so it might be worth revisiting.

I don't think it's possible to query the decorations height, but maybe that's something worth opening an issue over? The winit/glutin people probably know much more about what would be required to support this usecase. So sending a patch upstream might be the best solution for this.

I've thought about just deciding on a specific title bar height, however I feel like that's very prone to breakage. So we should try to check out what other options there are first.

**Comment by decrn** [comment]
I tracked down some of @sodiumjoe's work, including the proposed PR that you linked. 

Like you said @chrisduerr, I think this issue has more to do with winit than Alacritty itself, although I have read through some PRs and I get the feeling that the winit guys would really prefer to defer customizability downstream.

I'm not at all familiar with the codebase, and less so with macOS APIs, but as @sodiumjoe mentioned previously, we should be able to pull cocoa in and do this directly with `ns::get_window()`.

I've looked at [the documentation](https://developer.apple.com/documentation/appkit/nswindow) and it looks like we could fetch the default titlebar height using some minor _fuckery_ with the [`NSWindow.styleMask`](https://developer.apple.com/documentation/appkit/nswindow/stylemask) member. But this would hardly be useful as we need to know the title bar size ahead of time....

I also couldn't find anything on changing title bar heights throughout macOS revisions, as a way of comparing different versions.

To give you some perspective, gone are the days of iTunes traffic lights
![image](https://user-images.githubusercontent.com/15365152/59135903-050c7600-8981-11e9-830d-713e8d8c33a3.png)

I believe iTerm2's nightly build just adds a certain top padding, regardless of the macOS version etc. I suppose that's the lazy way out..

The [whole discussion](https://gitlab.com/gnachman/iterm2/issues/4080#note_156069487) is worth skimming over, including the arguments over 38 or 40 height (points) for the title bar to match Safari.

For comparison: Firefox offers "Normal", "Compact" and "Title bar" options:
<img width="106" alt="firefox" src="https://user-images.githubusercontent.com/15365152/59136135-05594100-8982-11e9-8857-f3152bc65acd.png">

**Comment by chrisduerr** [comment]
> I believe iTerm2's nightly build just adds a certain top padding, regardless of the macOS version etc. I suppose that's the easy way out..

Do you know where this is done in code? I'd be curious to see if they actually just pick a random value.

The Firefox comparison exposes something interesting though, it actually changes the decorations height, not just the top padding. As you can see, the padding above and below the window buttons gets bigger/smaller, which means that the decorations themselves have to know where they should be put.

If there would be a way to set the decorations height with transparent decorations, we could just use that to set it to some random value and be done with it.

**Comment by decrn** [comment]
> Do you know where this is done in code? I'd be curious to see if they actually just pick a random value.

I found [this referenced commit](https://gitlab.com/gnachman/iterm2/commit/41e75e20f) towards the end of the discussion:

So it seems they do just [pick a random value](https://gitlab.com/gnachman/iterm2/blob/master/sources/PseudoTerminal.m#L8096-8110).

> The Firefox comparison exposes something interesting though, it actually changes the decorations height, not just the top padding. As you can see, the padding above and below the window buttons gets bigger/smaller, which means that the decorations themselves have to know where they should be put.

This is something I also noticed compositing the image.


> If there would be a way to set the decorations height with transparent decorations, we could just use that to set it to some random value and be done with it.

It seems decorations can be generated using a given `styleMask`: https://developer.apple.com/documentation/appkit/nswindow/1419173-standardwindowbutton

I'm not so sure, maybe I'm skimming over it, but I think the easiest solution would be to draw a custom frame. I have no idea how to do this, though.

**Comment by jcf** [comment]
One small QOL improvement when configuring Marginalia is to add annotations to Projectile like so:

``` emacs-lisp
(add-to-list 'marginalia-prompt-categories '("Find file" . file))
(add-to-list 'marginalia-prompt-categories '("Switch to project" . file))
```
